### ISSUE TITLE
AppHost.OnEndRequest not executing under load (4.0.42)

### ISSUE THREAD

**Issue by rudism** [ServiceStack/Issues#329]
I'm overriding `OnEndRequest` in my AppHost, but we are seeing some cases where it does not get executed for some requests, possibly related to multipe requests coming in at the same time. I am using 4.0.42, it is a self-host project running on Linux using Mono 4.0.2.

I can reproduce this issue consistently use the following example app:

```cs
class MainClass
{
    public class AppHost : AppSelfHostBase
    {
        public AppHost() : base("testing", Assembly.GetExecutingAssembly()) { }

        public override void Configure(Container container)
        {
        }

        protected override Task ProcessRequestAsync(HttpListenerContext context)
        {
            Console.WriteLine("thread {0} begin", Thread.CurrentThread.ManagedThreadId);
            return base.ProcessRequestAsync(context);
        }

        public override void OnEndRequest(IRequest request = null)
        {
            Console.WriteLine("thread {0} end", Thread.CurrentThread.ManagedThreadId);
            base.OnEndRequest(request);
        }
    }

    public static void Main (string[] args)
    {
        new AppHost().Init().Start("http://*:9090/");
        "Running on port 9090".Print();
        Console.ReadLine();
    }

    [Route("/test")]
    public class TestMessage : IReturn<string>
    {
    }

    public class MainService : Service
    {
        public string Get(TestMessage request)
        {
            return string.Empty;
        }
    }
}
```
After running that project, doing a single request using `curl` works as expected, I see both the "begin" and "end" messages on the console where the server is running.

For example, the console output after running `curl -i localhost:9090/test` twice:

```
Running on port 9090
thread 9 begin
thread 9 end
thread 10 begin
thread 10 end
```

If I use `httperf` to blast the server with more connections in quick succession, only the "begin" messages print on the console. The "end" messages never show up. `httperf` reports no errors on any requests, but as far as I can tell `OnEndRequest` was never called.

The command I used was:

`httperf --server localhost --port 9090 --num-conns 100 --rate 10 --timeout 3 --uri "http://localhost:9090/test"`

That will open 10 connections per second for 10 seconds. The result was 100 `thread 11 begin` messages but not a single `thread 11 end` message.

Any idea what is going on here? Is there something I'm missing to ensure that our `OnEndRequest` override is always executed at the end of every request?

**Comment by rudism** [comment]
After further testing, it looks like it's not related to load at all, but rather `OnEndRequest` is not executing for requests that result in a 404. My `httperf` command was failing because the `--uri "http://localhost:9000/test"` switch was causing it to send that whole thing as the `GET` request. Changing it to `--uri "/test"` fixed it and caused all of the "end" messages to print.

My question remains, however. Shouldn't `OnEndRequest` execute on every request (even if they 404)? If not, is there an alternative method we can use to execute code at the end of every request?

**Comment by mythz** [comment]
Hi, executing Raw HttpHandlers are a way to short-circuit the [ServiceStack Request Pipeline](https://github.com/ServiceStack/ServiceStack/wiki/Order-of-Operations), i.e. they're not normal Service requests resolved from the IOC or managed by ServiceStack which wouldn't normally get notified if someone decides execute their own custom HttpHandler without explicitly calling `EndHttpHandlerRequest()` themselves, but I agree we should do this for any built-in custom HttpHandlers that we do have control over in ServiceStack which I've just [enabled in this commit](https://github.com/ServiceStack/ServiceStack/commit/4c9523aca80021babebc7ab48bef0eba99b88745).

This change is available from v4.0.43+ that's now [available on MyGet](https://github.com/ServiceStack/ServiceStack/wiki/MyGet).

**Comment by rudism** [comment]
I've just tested this using the MyGet version. One thing to note is that the `OnEndRequest` call is being made from a different thread, so I'm still not able to use this for what I wanted (cleaning up request/thread specific resources that are being allocated in the `ProcessRequestAsync` call). But maybe I can achieve what I want using request/response filters instead, I still have to figure it out.
